### ISSUE TITLE
Questions about the code.

### ISSUE THREAD

**Issue by haoz19** [facebookresearch/banmo#64]
Hello Gengshan,

Thanks for your contribution to video 3D reconstruction by introducing such a great approach and providing detailed instructions and code! Not only Banmo but also LASR, and ViSER all inspired me a lot.

I spent days trying to understand your code, which is already very clear, but I still have something unclear.

In the LASR code, you use 2 functions: obj_to_cam() and pinhole_cam() to project the vertexes from 3D root(canonical) coordinates to 2D coordinates. 
The LBS operation is within obj_to_cam(), which is from root space(rest pose) to view space.
pinhole_cam() is from camera 3D coordinates to 2D coordinatese.
![image](https://github.com/facebookresearch/banmo/assets/123277361/af6f6e58-732c-4443-b92c-534b7edddfb2)

And we want to do the same operation in banmo/nnutils/train_utils.py. We tried to follow the pipeline you provided: X(root space at rest pose) -> forward LBS -> X(root space at t') -> root pose -> X(view space at t'). 

And we try to use the forward lbs(bones, rts_fw, skin, xyz_in, backward=False), obj_to_cam(in_verts, Rmat, Tmat) and then pinhole_cam(in_verts, K). 
'K' could be obtained using 'self.model.rtk' and 'self.model.kaug'.
'Rmat' and 'Tmat' can be obtained by: root_rts = self.nerf_root_rts(dp_feats_rd).
'rts_fw' could be obtained by: bone_rts = self.nerf_body_rts(embedid).

But I'm stuck in the terms 'dp_feats_rd' and 'embedid', could you give us some insight about what those 2 terms stand for and how should we get those? Also is there anything I'm misunderstanding above?

Many Thanks!

Hao

**Comment by gengshan-y** [comment]
Hi Hao, thanks for the kind words. `dp_feats_rd` is the HxWx16 DensePose feature. `embedid` serves the same purpose as `frameid`, both of which are the frame index within the whole dataset.

So your understanding is mostly correct, except that you want to use `root_rts = self.nerf_root_rts(frameid)` to get object to view space transformations. Densepose features are only used together with the pre-trained CNN to initialize root poses, and never used during optimization.

**Comment by haoz19** [comment]
Hello Gengshan,

Thank you, We made some progress on projecting the vertices from rest space to 2D space. 

First, we get the rtk and kaug as follows:

rtk   = batch['rtk']         .view(bs,-1,4,4).permute(1,0,2,3).reshape(-1,4,4)    .to(self.device).float()
kaug   = batch['kaug']        .view(bs,-1,4).permute(1,0,2).reshape(-1,4)          .to(self.device).float()

Then we calculate Rmat, Tmat, K as follows:

Rmat = rtk[:,:3,:3]
Tmat = rtk[:,:3,3]
Kmat = K2mat(rtk[:,3,:])
Kaug = K2inv(kaug) # p = Kaug Kmat P
Kinv = Kmatinv(Kaug.matmul(Kmat))
K = mat2K(Kmatinv(Kinv))

Next, we get rts:

bs=batch['dataid'].shape[0]
data_info = self.init_dataset() 
data_offset = data_info['offset']
frameid      = batch['frameid']     .view(bs,-1).permute(1,0).reshape(-1).cpu()
dataid       = batch['dataid']      .view(bs,-1).permute(1,0).reshape(-1).cpu()
embedid = frameid + data_offset[dataid.long()]
frameid = frameid + data_offset[dataid.long()]
frameid = frameid.to(self.device)

bone_rts = self.model.module.nerf_body_rts(frameid)
rts_fw = correct_rest_pose(opts, bone_rts, bone_rts_rst)
rts_fw = rts_fw.view(-1,B,12)# B,12
rmat=rts_fw[:,:,:9]
rmat=rmat.view(-1,B,3,3)
tmat= rts_fw[:,:,9:12]
rts_fw = torch.cat([rmat,tmat[...,None]],-1)
rts_fw = rts_fw.view(-1,B,3,4)


And  use blend_skinning_chunk() -> obj_to_cam ->pinhole_cam 
to get the 2D projection. We get the following result

![scatter_plot_example (1)](https://github.com/facebookresearch/banmo/assets/123277361/dc485e98-f486-4866-8c19-0c5fdb1826f1)

We observed the result from blend_skinning_chunk seems reasonable however the projection result as shown in the image above is not in the right co-ordinate. (We think the right projection should start from (0,0) and up to (64,64) instead of  starting from some negative value in the image)

We guess the reason for the shift in the axis is because of the wrong K. Can you please advise on the right way to project the 3D to 2D? Additionally, can you please have a quick look at the codes we sent above and see if we are making any mistakes there?

Many Thanks!
